lib/on_form/form.rb

```diff
@@ -91,8 +90,21 @@ def self.expose_collection_of(association_name, on: nil, prefix: nil, suffix: ni
       collection_form_class.take_identity_from singular_name, convert_to_model: false
       collection_form_class.class_eval(&block)
 
-      define_method(exposed_name) { collection_wrappers[association_name] ||= CollectionWrapper.new(backing_model_instance(on), association_name, collection_form_class, allow_insert, allow_update, allow_destroy) } # used by action_view's fields_for, and by the following lines
+      define_method(exposed_name) do
+        return collection_wrappers.dig(on, association_name, :form) if collection_wrappers.dig(on, association_name)
+
+        collection_wrappers[on] ||= {}
+        collection_wrappers[on][association_name] = {
+          form: CollectionWrapper.new(backing_model_instance(on), association_name, collection_form_class, allow_insert, allow_update, allow_destroy), # used by action_view's fields_for, and by the following lines
+          exposed_name: exposed_name
+        }
+
+        collection_wrappers.dig(on, association_name, :form)
+      end
+
       define_method("#{exposed_name}_attributes=") { |params| send(exposed_name).parse_collection_attributes(params) }
+      define_method("_validate_#{exposed_name}_forms") { send(exposed_name).validate_forms }
+      after_validation :"_validate_#{exposed_name}_forms"
```

Comment:
Just wondering, since we have this callback, do you think we could/should add to the errors in that callback, instead of having to make the above change to add_collection_form_errors, which presumably meant you had to change the collection_wrappers stuff?

Reply:
Yep, That's a good idea. Unfortunately, we can't use after_validation. Instead of that, we have to use validate.